[Merged by Bors] - Recalculate entity aabbs when meshes change

mlodato517 opened the pull request:
# Objective

Update the `calculate_bounds` system to update `Aabb`s
for entities who've either:
- gotten a new mesh
- had their mesh mutated

Fixes https://github.com/bevyengine/bevy/issues/4294.

## Solution

There are two commits here to address the two issues above:

### Commit 1

**This Commit**
Updates the `calculate_bounds` system to operate not only on entities
without `Aabb`s but also on entities whose `Handle<Mesh>` has changed.

**Why?**
So if an entity gets a new mesh, its associated `Aabb` is properly
recalculated.

**Questions**
- This type is getting pretty gnarly - should I extract some types?
- This system is public - should I add some quick docs while I'm here?

### Commit 2

**This Commit**
Updates `calculate_bounds` to update `Aabb`s of entities whose meshes
have been directly mutated.

**Why?**
So if an entity's mesh gets updated, its associated `Aabb` is properly
recalculated.

**Questions**
- I think we should be using `ahash`. Do we want to do that with a
  direct `hashbrown` dependency or an `ahash` dependency that we
  configure the `HashMap` with?
- There is an edge case of duplicates with `Vec<Entity>` in the
  `HashMap`. If an entity gets its mesh handle changed and changed back
  again it'll be added to the list twice. Do we want to use a `HashSet`
  to avoid that? Or do a check in the list first (assuming iterating
  over the `Vec` is faster and this edge case is rare)?
- There is an edge case where, if an entity gets a new mesh handle and
  then its old mesh is updated, we'll update the entity's `Aabb` to the
  new geometry of the _old_ mesh. Do we want to remove items from the
  `Local<HashMap>` when handles change? Does the `Changed` event give us
  the old mesh handle? If not we might need to have a
  `HashMap<Entity, Handle<Mesh>>` or something so we can unlink entities
  from mesh handles when the handle changes.
- I did the `zip()` with the two `HashMap` gets assuming those would
  be faster than calculating the Aabb of the mesh (otherwise we could do
  `meshes.get(mesh_handle).and_then(Mesh::compute_aabb).zip(entity_mesh_map...)`
  or something). Is that assumption way off?

## Testing

I originally tried testing this with `bevy_mod_raycast` as mentioned in the
original issue but it seemed to work (maybe they are currently manually
updating the Aabbs?). I then tried doing it in 2D but it looks like
`Handle<Mesh>` is just for 3D. So I took [this example](https://github.com/bevyengine/bevy/blob/main/examples/3d/pbr.rs)
and added some systems to mutate/assign meshes:

<details>
<summary>Test Script</summary>

```rust
use bevy::prelude::*;
use bevy::render::camera::ScalingMode;
use bevy::render::primitives::Aabb;

/// Make sure we only mutate one mesh once.
#[derive(Eq, PartialEq, Clone, Debug, Default)]
struct MutateMeshState(bool);

/// Let's have a few global meshes that we can cycle between.
/// This way we can be assigned a new mesh, mutate the old one, and then get the old one assigned.
#[derive(Eq, PartialEq, Clone, Debug, Default)]
struct Meshes(Vec<Handle<Mesh>>);

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .init_resource::<MutateMeshState>()
        .init_resource::<Meshes>()
        .add_startup_system(setup)
        .add_system(assign_new_mesh)
        .add_system(show_aabbs.after(assign_new_mesh))
        .add_system(mutate_meshes.after(show_aabbs))
        .run();
}

fn setup(
    mut commands: Commands,
    mut meshes: ResMut<Assets<Mesh>>,
    mut global_meshes: ResMut<Meshes>,
    mut materials: ResMut<Assets<StandardMaterial>>,
) {
    let m1 = meshes.add(Mesh::from(shape::Icosphere::default()));
    let m2 = meshes.add(Mesh::from(shape::Icosphere {
        radius: 0.90,
        ..Default::default()
    }));
    let m3 = meshes.add(Mesh::from(shape::Icosphere {
        radius: 0.80,
        ..Default::default()
    }));
    global_meshes.0.push(m1.clone());
    global_meshes.0.push(m2);
    global_meshes.0.push(m3);

    // add entities to the world
    // sphere
    commands.spawn_bundle(PbrBundle {
        mesh: m1,
        material: materials.add(StandardMaterial {
            base_color: Color::hex("ffd891").unwrap(),
            ..default()
        }),
        ..default()
    });
    // new 3d camera
    commands.spawn_bundle(Camera3dBundle {
        projection: OrthographicProjection {
            scale: 3.0,
            scaling_mode: ScalingMode::FixedVertical(1.0),
            ..default()
        }
        .into(),
        ..default()
    });

    // old 3d camera
    // commands.spawn_bundle(OrthographicCameraBundle {
    //     transform: Transform::from_xyz(0.0, 0.0, 8.0).looking_at(Vec3::default(), Vec3::Y),
    //     orthographic_projection: OrthographicProjection {
    //         scale: 0.01,
    //         ..default()
    //     },
    //     ..OrthographicCameraBundle::new_3d()
    // });
}
fn show_aabbs(query: Query<(Entity, &Handle<Mesh>, &Aabb)>) {
    for thing in query.iter() {
        println!("{thing:?}");
    }
}

/// For testing the second part - mutating a mesh.
///
/// Without the fix we should see this mutate an old mesh and it affects the new mesh that the
/// entity currently has.
/// With the fix, the mutation doesn't affect anything until the entity is reassigned the old mesh.
fn mutate_meshes(
    mut meshes: ResMut<Assets<Mesh>>,
    time: Res<Time>,
    global_meshes: Res<Meshes>,
    mut mutate_mesh_state: ResMut<MutateMeshState>,
) {
    let mutated = mutate_mesh_state.0;
    if time.seconds_since_startup() > 4.5 && !mutated {
        println!("Mutating {:?}", global_meshes.0[0]);
        let m = meshes.get_mut(&global_meshes.0[0]).unwrap();
        let mut p = m.attribute(Mesh::ATTRIBUTE_POSITION).unwrap().clone();
        use bevy::render::mesh::VertexAttributeValues;
        match &mut p {
            VertexAttributeValues::Float32x3(v) => {
                v[0] = [10.0, 10.0, 10.0];
            }
            _ => unreachable!(),
        }
        m.insert_attribute(Mesh::ATTRIBUTE_POSITION, p);
        mutate_mesh_state.0 = true;
    }
}

/// For testing the first part - assigning a new handle.
fn assign_new_mesh(
    mut query: Query<&mut Handle<Mesh>, With<Aabb>>,
    time: Res<Time>,
    global_meshes: Res<Meshes>,
) {
    let s = time.seconds_since_startup() as usize;
    let idx = s % global_meshes.0.len();
    for mut handle in query.iter_mut() {
        *handle = global_meshes.0[idx].clone_weak();
    }
}
```

</details>

## Changelog

### Fixed
Entity `Aabb`s not updating when meshes are mutated or re-assigned.

mlodato517 commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -121,15 +123,47 @@ impl Plugin for VisibilityPlugin {
     mut commands: Commands,
     meshes: Res<Assets<Mesh>>,
-    without_aabb: Query<(Entity, &Handle<Mesh>), (Without<Aabb>, Without<NoFrustumCulling>)>,
+    without_aabb_or_with_changed_mesh: Query<
+        (Entity, &Handle<Mesh>),
+        (
+            Or<(Without<Aabb>, Changed<Handle<Mesh>>)>,
+            Without<NoFrustumCulling>,
+        ),
+    >,
```
This `Query` type is too complex for `clippy`'s likes. It doesn't seem trivial to extract but we could try something like:
```rust
type NoAaabOrMeshChanged = Or<(Without<Aabb>, Changed<Handle<Mesh>>)>;
type NeedsAabbUpdate = (NoAaabOrMeshChanged, Without<NoFrustumCulling>);
```
and then do:
```suggestion
    without_aabb_or_with_changed_mesh: Query<
        (Entity, &Handle<Mesh>),
        NeedsAabbUpdate,
    >,
```
but it didn't seem _obviously_ better :shrug:

alice-i-cecile replied:
Agreed, just leave it. We have the type_complexity lint turned off at a project level already.

You can run (most of) CI locally using `cargo run -p ci`.

mlodato517 commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -121,15 +123,47 @@ impl Plugin for VisibilityPlugin {
+    for (entity, mesh_handle) in without_aabb_or_with_changed_mesh.iter() {
+        // Record entities that have mesh handles.
+        entity_mesh_map
+            .entry(mesh_handle.clone())
```
How cheap is cloning the `mesh_handle`? If cloning is more expensive than hashing we could presumably do a `contains`+`insert` check instead of `entry` :balance_scale:

alice-i-cecile replied:
Extremely cheap: they're basically just a `Rc` smart pointer.

mockersf replied:
yes!
`clone` is cheap but is reference counted, `clone_weak` isn't so it's quite cheaper

mlodato517 replied:
Updated to `clone_weak` in 92e78bc2d90c09db4fc4502efa2d8ae22c311753.

mlodato517 commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -121,15 +123,47 @@ impl Plugin for VisibilityPlugin {
+    let updated_meshes_and_entities = updated_mesh_handles.filter_map(|mesh_handle| {
+        meshes
+            .get(mesh_handle)
+            .zip(entity_mesh_map.get(mesh_handle))
```
This list never gets entities removed from it. So if an entity gets a new handle and then its old handle is mutated we'd set its Aabb to the new geometry of the old mesh. We probably don't want that. Is there an easier way to keep track of these than another `Local<HashMap<Entity, Handle<Mesh>>>`?

alice-i-cecile replied:
Ah. This is https://github.com/bevyengine/bevy/issues/4513. I would go with the naive uncached version, and leave a `PERF` comment.

As you've seen, there's a lot of challenges regarding data synchronization when trying to roll your own indexes.

mlodato517 replied:
Attempted to handle in 5ee488a5e60ff51e30f309b103c6a4f9a12b1606. Updated PR description with an app that maybe shows the bug and the fix.

alice-i-cecile commented:
This is a fantastic contribution; thank you!

> This type is getting pretty gnarly - should I extract some types?

I don't think so; it's not reused, and it'll result in indirections. You get good at reading query signatures with time. My main gripe with it is the double negative `Without<NoFrustumCulling>`, but that's not your fault.

> This system is public - should I add some quick docs while I'm here?

Yes please!

> I think we should be using ahash. Do we want to do that with a
direct hashbrown dependency or an ahash dependency that we
configure the HashMap with?

`bevy_utils::HashMap` does exactly this; use it instead.

> There is an edge case of duplicates with `Vec<Entity>` in the
  `HashMap`. If an entity gets its mesh handle changed and changed back
  again it'll be added to the list twice. Do we want to use a `HashSet`
  to avoid that? Or do a check in the list first (assuming iterating
  over the `Vec` is faster and this edge case is rare)?

I would expect this edge case to be very rare, and the costs of using the `HashSet` non-trivial when we're modifying large numbers of mesh handles at once. IMO we should just repeat the work.

Please leave a comment to this effect :)

> There is an edge case where, if an entity gets a new mesh handle and
  then its old mesh is updated, we'll update the entity's `Aabb` to the
  new geometry of the _old_ mesh. Do we want to remove items from the
  `Local<HashMap>` when handles change? Does the `Changed` event give us
  the old mesh handle? If not we might need to have a
  `HashMap<Entity, Handle<Mesh>>` or something so we can unlink entities
  from mesh handles when the handle changes.
  
  The `Changed` information won't cache the old data. I'll need to chew on what should be done here.

> I did the `zip()` with the two `HashMap` gets assuming those would
  be faster than calculating the Aabb of the mesh (otherwise we could do
  `meshes.get(mesh_handle).and_then(Mesh::compute_aabb).zip(entity_mesh_map...)`
  or something). Is that assumption way off?
  
  I have no good intuition for this particular question.

alice-i-cecile commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -1,10 +1,12 @@
 mod render_layers;
 
+use std::collections::HashMap;
```
```suggestion
use bevy_utils::HashMap;
```

mlodato517 replied:
Updated in d266354f282c2943679144f8c0017596fc28e96d.

alice-i-cecile commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -121,15 +123,47 @@ impl Plugin for VisibilityPlugin {
+        entity_mesh_map
+            .entry(mesh_handle.clone())
+            .or_default()
+            .push(entity);
```
Like I said in the main thread, I think this is the right tradeoff. We should leave a comment documenting the edge case though.

mlodato517 replied:
Documented in 7b20f607f6eba53fcc3a417f691d167b90e22f44.

mlodato517 replied:
Note that in b02be5c18b8f29a6ac8aac6bda3dc466dffed11a I deleted the comment. I believe that, the way `register` and `deregister` are used, this is no longer possible. It may even be better if I move `deregister` to be the first thing before `register` so it can't be used wrong (but would also involve needless calls to `deregister` when registering an entity<->mesh for the first time). What do y'all think?

mlodato517 replied:
Also, if it _is_ still possible, then we probably want `deregister` to do a `while let Some(idx) = entities.position()` so that we delete all occurrences.

nicopap commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -121,15 +123,47 @@ impl Plugin for VisibilityPlugin {
     }
+
+    // Calculate bounds for entities whose meshes have been mutated.
+    let updated_mesh_handles = mesh_events.iter().filter_map(|event| match event {
+        AssetEvent::Modified { handle } => Some(handle),
+        _ => None,
+    });
+    let updated_meshes_and_entities = updated_mesh_handles.filter_map(|mesh_handle| {
+        meshes
+            .get(mesh_handle)
+            .zip(entity_mesh_map.get(mesh_handle))
+    });
+    for (mesh, entities) in updated_meshes_and_entities {
+        if let Some(aabb) = mesh.compute_aabb() {
+            for entity in entities {
+                commands.entity(*entity).insert(aabb.clone());
+            }
+        }
+    }
```
`zip` is fine here, but I'd suggest the following code.
```rust
let to_update = |event| {
    let handle = match event {
        AssetEvent::Modified { handle } => handle,
        _ => return None,   
     };
    let mesh = meshes.get(handle)?;
    let with_handle = entity_mesh_map.get(handle)?;
    let aabb = mesh.compute_aabb()?;
    Some((aabb, with_handle))
};
for (aabb, with_handle) in events.iter().filter_map(to_update) {
    for to_update in with_handle {
        commands.entity(to_update).insert(aabb.clone());
    }
}```

mlodato517 replied:
Awesome! Updated in 079a9930600907484085b0b2c1631f4a4b54751b.

nicopap reviewed (changes requested):
Regarding the question on the edge case. I think you are right. We'd need to keep a `HashMap<Entity, Handle<Mesh>>`. So that it's possible to remove the `Entity` from right entry when things gets updated or removed (I think it might be less complex and more performant this way) (also check `RemovedComponents` for the case when an entity is despawned or have their `Handle<Mesh>` removed).

You can also have a `update_bounds` systems that update in-place the `Aabb`s with a `Query<&mut Aabb>`, this way, the update logic is separated from the create logic and the `Aabb`s are updated as soon as the `update_bounds` system is complete. but I don't think it's a big deal.

mlodato517 commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -1,10 +1,12 @@
 mod render_layers;
 
+use std::collections::HashMap;
```
Awesome, thanks! I'll resolve this as a duplicate of https://github.com/bevyengine/bevy/pull/4944#discussion_r890196927.

mlodato517 commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -121,15 +123,47 @@ impl Plugin for VisibilityPlugin {
+    for (entity, mesh_handle) in without_aabb_or_with_changed_mesh.iter() {
+        // Record entities that have mesh handles.
+        entity_mesh_map
+            .entry(mesh_handle.clone())
```
Great! I'll resolve this as a duplicate of https://github.com/bevyengine/bevy/pull/4944#discussion_r890221084.

superdump commented:
As the original author of the frustum culling code, I’d like to review this when I have time.

mlodato517 commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -118,18 +121,68 @@ impl Plugin for VisibilityPlugin {
+/// # Examples
+///
+/// To automatically get calculated `Aabb`s that will update with `Mesh` assignment/mutation, add
+/// this as a [`System`] to your `App`:
```
I couldn't get the link to work for `App` and I have no idea why.

mlodato517 replied:
Removed in 9730917d7a5ea6082b8966f5f6e3dd12724d7317.

superdump commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -118,18 +121,68 @@ impl Plugin for VisibilityPlugin {
+/// # use bevy_render::view::calculate_bounds;
+///
+/// App::new()
+///     .add_system(calculate_bounds)
```
Do we normally do this kind of example for a core system that is part of the default systems that are enabled? It seems a bit overboard to me as this is how adding any system to a bevy app works.

mlodato517 replied:
Oh I didn't know this was a default system! After reading the original issue I was a bit interested in how users were supposed to discover this. If it's a default system then I'd also vote to remove this example (though I'll defer to anyone else here).

alice-i-cecile replied:
Yep, I don't think describing how to add default systems like this is useful.

mlodato517 replied:
Removed in 9730917d7a5ea6082b8966f5f6e3dd12724d7317

superdump commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -118,18 +121,68 @@ impl Plugin for VisibilityPlugin {
 }
 
+/// Calculates and updates [`Aabb`]s for [`Entities`](Entity) with [`Mesh`]es.
+/// To opt out of bound calculation for an `Entity`, give it the [`NoFrustumCulling`] component.
```
Keep these ^ two lines of documentation of the system, but I don't feel the rest is useful personally.

mlodato517 replied:
Maintained in 9730917d7a5ea6082b8966f5f6e3dd12724d7317.

superdump reviewed (commented):
Pretty much looks good to me, just a comment about what I perceive to be unnecessary documentation. Maybe @alice-i-cecile disagrees?

mlodato517 commented:
> We'd need to keep a `HashMap<Entity, Handle<Mesh>>`. So that it's possible to remove the `Entity` from right entry when things gets updated or removed (I think it might be less complex and more performant this way) (also check `RemovedComponents` for the case when an entity is despawned or have their `Handle<Mesh>` removed).
> 
> You can also have a `update_bounds` systems that update in-place the `Aabb`s with a `Query<&mut Aabb>`, this way, the update logic is separated from the create logic and the `Aabb`s are updated as soon as the `update_bounds` system is complete. but I don't think it's a big deal.

I'm a little confused here. I like the idea of the separate `update_bounds` system because it sounds much cleaner but it won't have access to `calculate_bound`'s `Local<HashMap>`. If we want to move the entire `Local<HashMap>` to `update_bounds` then I think it'll be hard to tell when to add an entity to a mesh handle's bucket.

Here's the pseudocode I was imagining - lemme know if that was the wrong read:
<details>
<summary>pseudocode</summary>

```rust
fn update_bounds(
    meshes: Res<Assets<Mesh>>,
    mesh_events: EventReader<AssetEvent<Mesh>>,
    mut to_update: Query<
        (Entity, &mut Aabb, &Handle<Mesh>),
        With<Changed<Handle<Mesh>>>,
    >,
    mut entities_with_mesh: Local<HashMap<Mesh, Entities>>,
    mut mesh_for_entity: Local<HashMap<Entity, Mesh>>,
    mut entities_with_removed_mesh: RemovedComponents<Handle<Mesh>>,
) {
    // If an entity no longer has a mesh, remove it from the list
    // of entities with that mesh.
    for entity in entities_with_removed_mesh {
        let mesh = &mesh_for_entity[entity];
        let entities_with_mesh = &mut entities_with_mesh[mesh];
        entities_with_mesh.swap_remove(get_idx(entity));
    }

    // If a mesh was mutated, update the Aabb for all entities with that mesh
    for event in mesh_events {
        let entities_with_mesh = &entities_with_mesh[event.mesh];
        let mesh = &meshes[event.mesh];
        let aabb = mesh.aabb();
        entities_with_mesh.iter_mut().for_each(|e| e.aabb = aabb.clone());
    }

    // Update Aabb for entities that got a new mesh
    for (entity, &mut aabb, mesh) in to_update {
        let mesh = meshes[mesh];
        let new_aabb = mesh.aabb();
        *aabb = new_aabb;

        // Does it go here? Does `Changed` work when things are added
        // for the first time? If not we'll probably fail on some mesh mutation
        // events. Maybe we can use `ChangeTrackers` to get changed _and_ added?
        entities_with_mesh.insert(mesh, entity);
    }
}
```

</details>

nicopap commented:
@mlodato517 You would need to have the `HashMap<Mesh, Entities>` and `HashMap<Entity, Mesh>` in a `Res` shared between the two systems, probably create a new struct and have it inserted as a resource.

mlodato517 commented on crates/bevy_render/src/view/visibility/mod.rs:
````diff
@@ -118,18 +121,68 @@ impl Plugin for VisibilityPlugin {
+/// ```
+/// # use bevy_app::prelude::App;
+/// # use bevy_render::view::calculate_bounds;
+///
````
Removed in 9730917d7a5ea6082b8966f5f6e3dd12724d7317.

mlodato517 commented:
Added the maps in a resource in 5ee488a5e60ff51e30f309b103c6a4f9a12b1606 and added an `update_bounds` system. Hopefully this is somewhere along the lines of what y'all were thinking :crossed_fingers:

mlodato517 commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -68,9 +71,44 @@ impl VisibleEntities {
+            .entry(mesh_handle.clone_weak())
+            .or_default()
+            .push(entity);
+        self.mesh_for_entity
```
In theory this `insert` should never return anything. Should we do anything if something _does_ come back out?

mlodato517 commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -83,47 +121,55 @@ impl Plugin for VisibilityPlugin {
     fn build(&self, app: &mut bevy_app::App) {
         use VisibilitySystems::*;
 
-        app.add_system_to_stage(
```
Best viewed with "hide whitespace" - added a resource and system and it changed the indentation

mlodato517 commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -68,9 +71,44 @@ impl VisibleEntities {
+}
+
+impl EntityMeshRelationships {
+    /// Register the passed `entity` has having the passed `mesh_handle`.
```
```suggestion
    /// Register the passed `entity` as having the passed `mesh_handle`.
```

mlodato517 replied:
Updated in 6061951588c4b006783c58d302ab732d639ff919

mlodato517 commented:
Huh the rerequest button doesn’t seem to be working for nicopap 🤷

alice-i-cecile commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -68,9 +71,44 @@ impl VisibleEntities {
 }
 
+#[derive(Debug, Default, Clone)]
+pub struct EntityMeshRelationships {
```
Docs please; this is a public struct.

nicopap replied:
`#[doc(hidden)]` could also work. Since it's an internal struct that really only is exposed because it's a parameter to a public system, and is not intended to be used outside of this module.

I'd also remove all the derives and use an `insert_resource(EntityMeshRelationships {.. })` so that it's literally impossible for the end-user to interact with this type.

alice-i-cecile replied:
Docs are useful to contributors too though :) And I think you'd be surprised at the sort of things users might want to do.

Making the system public without exposing a constructor for the resource is completely useless.

nicopap replied:
Fair enough. Though it's still not much use if the only thing the user can do with it is resetting the value. Which will only break the internals.

mlodato517 replied:
> Making the system public without exposing a constructor for the resource is completely useless.

Just checking - are you asking I implement a `new()` for this or are you just referencing the derived `Default::default` as the constructor and so you want to keep that in?

mlodato517 replied:
Added some documentation in bca16da88fa175a7dc7a89fcd74085fe303f69e8.

IceSentry replied:
I'm not sure I understand the point of this struct. Why is Query<(Entity, Mesh)> not enough information?

IceSentry replied:
I'm not sure I understand the point of this struct. Why is Query<(Entity, Mesh)> not enough information?

IceSentry replied:
It seems like it's needed because `AssetEvent<Mesh>` doesn't tell you which entities are affected by the change. I think it's important to document that because it wasn't obvious to me at all.

I originally assumed the implementation would just be a system with `Query<(Entity, Handle<Mesh>), Changed<Mesh>>` and it's really not obvious why that's not enough.

alice-i-cecile replied:
> It seems like it's needed because AssetEvent<Mesh> doesn't tell you which entities are affected by the change.

Could we just change this? This seems like a common need.

IceSentry replied:
Yah, it could make sense to handle that correctly in bevy_asset, but I don't want to block this PR on that. I think it's fine the way it is, it just needs more documentation to actually explain what's going on.

mlodato517 replied:
Updated the docs in b0f85d31dd73a7fd9d02cf12b093f6b48ea7df8d.

james7132 commented:
It's not something I want us to address this, but skinned meshes are an extreme example of this, where the AABB both out of date and inaccessible. We probably will need an exception for skinned meshes, and precompute a "extremes" bound for them using asset preprocessing of some kind.

nicopap commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -68,9 +71,44 @@ impl VisibleEntities {
+
+    /// Deregisters the relationship between an `Entity` and `Mesh`. Used so [`update_bounds`] can
+    /// track which relationships are still active so `Aabb`s are updated correctly.
+    fn deregister(&mut self, entity: Entity) {
```
Yeah. Sucks that the runtime is proportional to the number of entities with a `Handle<Mesh>` that gets updated or removed, so that removal of large quantities, or updating a single mesh that is shared across many entities. I don't think it's possible to get around it though. I'd suggest documenting that in the `update_bounds` system. It's important that users are aware of cost and benefits of various functionalities.

mlodato517 replied:
Documented in f9df58adecf5b03d1d6c6845eeda69b219ac65bd - lemme know if that makes enough sense!

nicopap commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -68,9 +71,44 @@ impl VisibleEntities {
+
+    /// Deregisters the relationship between an `Entity` and `Mesh`. Used so [`update_bounds`] can
+    /// track which relationships are still active so `Aabb`s are updated correctly.
+    fn deregister(&mut self, entity: Entity) {
+        if let Some(mesh) = self.mesh_for_entity.remove(&entity) {
+            if let Some(entities) = self.entities_with_mesh.get_mut(&mesh) {
+                if let Some(idx) = entities.iter().position(|&e| e == entity) {
+                    entities.swap_remove(idx);
+                }
+            }
+        }
+    }
```
Avoid rightward drift by using the `?` operator.
```suggestion
    fn deregister(&mut self, entity: Entity) -> Option<()> {
        let mesh = self.mesh_for_entity.remove(&entity)?;
        let entities = self.entities_with_mesh.get_mut(&mesh)?;
        let idx = entities.iter().position(|&e| e == entity)?;
        entities.swap_remove(idx);
        Some(())
    }
```

mlodato517 replied:
Hmmm I don't love adding a not-so-meaningful return value to this function just to get that syntax :thinking: That being said, [ties always go to the reviewer](https://mtlynch.io/code-review-love/#12-award-all-ties-to-your-reviewer) :-) (might not even be a tie since we have a thumbs up too!)

I'll propose one alternative but I'm happy to go with `?` if we still prefer that!:
```rust
    fn deregister(&mut self, entity: usize) {
        if let Some(entities) = self
            .mesh_for_entity
            .remove(&entity)
            .and_then(|mesh| self.entities_with_mesh.get_mut(&mesh))
        {
            if let Some(idx) = entities.iter().position(|&e| e == entity) {
                entities.swap_remove(idx);
            }
        }
    }
```

mlodato517 replied:
Yeah I don't know if I want `Option<()>` for "did this remove anything". I'd either choose a `bool` or `Result` (if it's an error for nothing to have happened which might be this case?) :thinking:

We could return the thing removed by `swap_remove` to be a little less janky but no consumer ever cares because it'll always be the same thing as what you passed in :shrug: 

Having the closure seems a little weird to me too but maybe just because it's not my habit to explicitly write closures :sweat_smile: I'll ponder this for a bit and then probably just give up and do one of the things you've suggested :laughing:

mlodato517 replied:
Actually I've had one more thought that I wonder if you're okay with. This `EntityMeshRelationships` maintains an invariant by having its `register` method add to two collections and its `deregister` method remove from those same two collections. I propose:
1. keeping the first lookup into `self.mesh_for_entity` as an `Option` that might be `None` because we have no idea what `Entity` was just passed to us
2. having the followup `get_mut` be a `&mut self.entities_with_mesh[&mesh]` because this _should not fail_ - if there was a mesh for the entity in `self.mesh_for_entity` there should _absolutely_ be a list of entities for that mesh
3. having the `position()` check end with an `.expect` (or similar) for the same reason - if we had a mesh in `self.mesh_for_entity` then that entity also should've been in the list for its mesh in `self.entities_with_mesh`

I _slightly_ don't like this because now you have to trust the invariants to make sure it doesn't crash and you have to use the provided methods (which we could make easier by sticking it in a module) but I think I might prefer it to the position of "this can't happen but I'm checking for it anyway because I'm not confident in my system".

We could slightly improve this by having `mesh_for_entity` not store a Mesh but instead a pointer to the `Entity` in `entities_for_mesh` but I'm not sure if that's quite worth it yet.

What do you think? Did I miss a possible edge case where this could fail or do we just not want to do this out of an abundance of caution?

nicopap replied:
It's up to you. I believe library code should panic as little as possible. It's not cool to break user code because of a small bug in your own code (I've been responsible for breaking user code a few time and it feels bad). If you feel confident in your own code, well, it's fine to use those panicking methods. But you also have to take into consideration bevy is built by a... bevy of people, and is in active development. So things are bound to change and invariant assumptions may break as a result. And you wouldn't want to panic if a bug is introduced in the future, a `warn` or `error` log should do :)

I would recommend against `unsafe` since it puts you on "cart approval track" by marking the PR automatically as controversial.

mlodato517 replied:
> bevy is built by a... bevy of people

This deserves a :1st_place_medal: 

> invariant assumptions may break as a result

Yeah this is something I hate to acknowledge. I'd hate to think that, when working on `BinaryHeap`, I'd have to worry that it's invariants would be broken. That being said, I'm not so confident to assert that here since this is more complicated than `BinaryHeap` for sure!

Now I just need to figure out which not-rightward-drifting code I like the most :-)

Also do you have any preference for the `if let Some(idx)` vs `while let Some(idx)` which I just thought of in https://github.com/bevyengine/bevy/pull/4944#discussion_r895075123?

mlodato517 replied:
I chose the closure syntax :-) e6c1a1f82308ba603e5fef545c59649d076dcf74

nicopap commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -132,6 +178,61 @@ pub fn calculate_bounds(
+        entity_mesh_rel.deregister(entity);
+    }
+
+    // If an entity with an Aabb was assigned a new mesh, move it to the new list and update its
+    // Aabb.
+    for (entity, mesh_handle, mut aabb) in mesh_reassigned.iter_mut() {
+        // Remove it from the previous list
+        entity_mesh_rel.deregister(entity);
+
+        // Add it to the new list.
+        entity_mesh_rel.register(entity, mesh_handle);
+
+        // Update its Aabb
+        if let Some(mesh) = meshes.get(mesh_handle) {
+            if let Some(new_aabb) = mesh.compute_aabb() {
+                *aabb = new_aabb;
+            }
+        }
+    }
```
```suggestion
    for (entity, mesh_handle, mut aabb) in mesh_reassigned.iter_mut() {
        entity_mesh_rel.deregister(entity);
        entity_mesh_rel.register(entity, mesh_handle);
        if let Some(mesh) = meshes.get(mesh_handle) {
            if let Some(new_aabb) = mesh.compute_aabb() {
                *aabb = new_aabb;
            }
        }
    }
```

Please don't use comments that describe **what** the code does. You should only write comments if it's unclear **why** it does something. The code is **what** the code does, comments that tell **what** the code does are redundant.

mlodato517 replied:
Ah, good catch. I 100% agree.

mlodato517 replied:
Comments removed in 76197ccee336f6f073038959b11d19f740333243.

nicopap reviewed (changes requested):
Looks good to me. Fixes #4294 alright. I think the comments on documentation should be addressed and then it's good for me.

nicopap commented:
> It's not something I want us to address in this PR, but skinned meshes are an extreme example of this, where the AABB both out of date and inaccessible basically every frame. We probably will need an exception for skinned meshes, and precompute a "extremes" bound for them using asset preprocessing of some kind.

@james7132  I opened an issue so that it's possible to discuss this independently in #4971

nicopap commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -83,47 +121,55 @@ impl Plugin for VisibilityPlugin {
+    mut entity_mesh_rel: ResMut<EntityMeshRelationships>,
 ) {
     for (entity, mesh_handle) in without_aabb.iter() {
+        // Record entities whose Aabb's we're managing.
+        entity_mesh_rel.register(entity, mesh_handle);
+
         if let Some(mesh) = meshes.get(mesh_handle) {
             if let Some(aabb) = mesh.compute_aabb() {
                 commands.entity(entity).insert(aabb);
```
```suggestion
        if let Some(mesh) = meshes.get(mesh_handle) {
            if let Some(aabb) = mesh.compute_aabb() {
                entity_mesh_rel.register(entity, mesh_handle);
                commands.entity(entity).insert(aabb);
```
the call to `register` should probably be added here, so that it mirrors the actual state of the ECs if one of the `if let` fails.

mlodato517 replied:
Oh actually ... hmmm ... I'm not so sure. I have two competing thoughts.

**Why It Should Be Inside the If Lets**
If an entity has a mesh and does not have an Aabb then it'll be in this query. If the assets collection doesn't have a mesh for the entity's handle or the Aabb can't be computed then this entity will be here the next time this system is called. We don't want to add it to the relationship map twice.

**Why It Should Be Outside the If Lets**
`entity_mesh_rel` tracks which entities have which meshes. If we fail to compute an Aabb for a mesh that doesn't mean that the entity _doesn't_ have that mesh. So we should record it regardless

That being said I'm leaning towards doing what you suggest but it might help if I knew more about the failure states of `meshes.get` and `compute_aabb` :thinking:

mlodato517 replied:
Actually yeah I guess it's good to not add things until we compute an `Aabb` for it. If an entity has a mesh but not an Aabb and then its mesh is mutated we probably don't want to try to assign a mesh _and_ try and update the mesh :thinking:

mlodato517 replied:
Updated in b02be5c18b8f29a6ac8aac6bda3dc466dffed11a.

nicopap replied:
`meshes.get()` is likely to fail at least once. `Handle<Mesh>`s are added to entities before the mesh is fully loaded, and it's possible that a few frames before the mesh gets loaded that it returns `None`.

`compute_aabb` should only fail if the mesh doesn't have `ATTRIBUTE_POSITION` or has malformed vertex positions. bevy supports custom mesh attributes, so I think it's possible some meshes will have their `compute_aabb` fail all the time. The more I think about this, the more I believe the `aabb` should be part of the `Mesh`. But that's for another time.

nicopap commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -132,6 +178,61 @@ pub fn calculate_bounds(
+    // Aabb.
+    for (entity, mesh_handle, mut aabb) in mesh_reassigned.iter_mut() {
+        // Remove it from the previous list
+        entity_mesh_rel.deregister(entity);
+
+        // Add it to the new list.
+        entity_mesh_rel.register(entity, mesh_handle);
+
+        // Update its Aabb
+        if let Some(mesh) = meshes.get(mesh_handle) {
+            if let Some(new_aabb) = mesh.compute_aabb() {
+                *aabb = new_aabb;
```
So would be a good idea to move the `register` and `deregister` inside the `if let` block here as well.

mlodato517 replied:
Updated in b02be5c18b8f29a6ac8aac6bda3dc466dffed11a. Note that we _always_ deregister the entity<->mesh. I think that makes sense.

mlodato517 commented:
What's the usual procedure for resolving comments? I was going to let reviewers resolve comments so they know I've done what they've asked correctly. If y'all'd prefer I resolve the comments that I think I've handled I can do that too and clean up the thread!

alice-i-cecile commented:
Only people with write permissions (aka Cart) and PR authors can resolve comments. As an author I typically leave them for a bit and then resolve them myself.

nicopap reviewed (approved):
I think this is ready to be merged.

alice-i-cecile commented:
bors try

IceSentry commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -83,55 +119,112 @@ impl Plugin for VisibilityPlugin {
 }
 
+/// Calculates [`Aabb`]s for [`Entities`](Entity) with [`Mesh`]es. To opt out of bound calculation
+/// for an `Entity`, give it the [`NoFrustumCulling`] component.
```
I think we should have another way to opt out of that. If, for example, a mesh is updated often, but the actual bounds of the mesh don't change then opting out of frustum culling isn't ideal. You'd still want to cull the mesh if it's not visible, you just don't want it to re calculate the bounds constantly.

nicopap replied:
I agree with this. Adding a marker component like `NoAabbCompute` would be great.

mlodato517 replied:
I imagine we want to create this new marker trait and:
1. add it to `calculate_bounds`
2. add it to `update_bounds`

Do we also want to remove `NoFrustumCulling` from `update_bounds`? Or should we have both?

IceSentry replied:
I think it should keep `NoFrustumCulling`. It makes sense to not compute the bounds with that because it becomes unnecessary.

mlodato517 replied:
Hopefully e75a7bf593718c1b1c7a2da695cf84cbf8fa910f does this right :crossed_fingers:

IceSentry commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -83,55 +123,123 @@ impl Plugin for VisibilityPlugin {
+        (
+            Without<Aabb>,
+            Without<NoFrustumCulling>,
+            Without<NoAabbCompute>,
```
I don't think this should be here. It makes sense to compute it at least once, I think it's just recomputing the boundsthat should be avoided with this tag.

IceSentry replied:
I think it's overkill yes, as you said nobody has complained about having something that can't be calculated. A lot of people have had issues with it not updating automatically though.

The reason I mentioned that marker originally is for situations where you want frustum culling and you know you will update a mesh frequently but the overall size of the mesh won't change. This is technically an optimization that might not be needed though, but I think it makes sense to only have `NoAabbRecompute` (I'd suggest calling it `NoAabbUpdate`, but that doesn't really matter). 

If people don't want any Aabb, then they probably don't want FrustumCulling so just having the `Without<NoFrustumCulling>` should be enough.

mlodato517 replied:
Renamed to `NoAabbUpdate` and removed it from `calculate_bounds` in 7ac4fec6135b9a4d3ffa83b004fb9b3e889653ba.

mlodato517 commented:
Just did a rebase in case that was why CI failed. Should've checked locally before pushing just to make sure...

IceSentry reviewed (approved):
With the docs I'm now happy with this. It's unfortunate that we need to keep track of the relations for mesh assets, but this is still a super useful feature and worth the complexity until we fix it in the asset pipeline.

(of course, CI still needs to pass, but I'm happy with the design)

mlodato517 commented:
Thanks! Before I clean this up to merge I want to resolve a question about future-proofing against a particular kind of ABA bug. I think my questions got lost in the [individual](https://github.com/bevyengine/bevy/pull/4944#discussion_r895075803) [comments](https://github.com/bevyengine/bevy/pull/4944#discussion_r894922911) so I'll drop it here.

Currently we only `register` an entity<->mesh mapping if:
1. we're seeing this entity-with-mesh for the first time
2. this entity has a `Changed<Handle<Mesh>>` (i.e. the entity got a new mesh)

Currently we `deregister` an entity<->mesh mapping if:
1. the entity's mesh was removed (as reported by `RemovedComponents`)
2. the entity has a `Changed<Handle<Mesh>>` which means it was reassigned which means it no longer has its old mesh

Currently everything is good (I think). However, if something were to change in these systems and it was possible to call `register` on an entity<->mesh combination twice without `deregister`ing the first mapping then you could end up in a situation where:
1. you register Entity A with Mesh B twice
2. Entity A gets reassigned Mesh C (so one A<->B mapping is removed but the other remains)
3. Mesh B gets mutated so we update every entity associated with Mesh B which includes Entity A
4. Entity A now has an `Aabb` for Mesh B even though it's currently using Mesh C

I can see at least a few easy ways forward:
1. Don't change any code and maybe document this behavior so future devs are careful
2. Update the collection to be a `HashSet` so there can't be duplicates (but iteration is slower - this idea was [already discussed](https://github.com/bevyengine/bevy/pull/4944#issuecomment-1147498636) and I think we want to keep it off the table but I'm mentioning it here for completeness)
3. Update `deregister` to remove mappings in a loop or use `retain` to ensure that all relationships for the given entity are removed
    1. a loop with `swap_remove` is a bit of a bummer because the common case of 1 relationship would always incur two sweeps - one to remove the first relationship and 1 to see that there are no more
    2. `retain` would work seamlessly with 1 or more relationships but in the common case of 1 relationship will wastefully move entities around to fill gaps
4. Call `deregister` as the first thing in `register` so we always register fresh relationships - this will cause lots of no-op `HashMap.remove()`s
5. Check the list first before registering an entity to make sure it isn't already there (this was also discussed earlier and we probably don't want to do that)

Does anyone have guidance/a strong opinion here?

mlodato517 commented:
> of course, CI still needs to pass

Yeah this might need to wait for a rebase. It looks like CI is failing on `main` too. `cargo run -p ci` passes for me locally but it looks like hosted CI is failing for some deprecated serializer in `bevy_reflect`.

alice-i-cecile commented:
Merging #5021 as we speak to fix the failing CI.

mlodato517 commented:
I'm glad CI passed. One question - when I run `cargo run -p ci` locally I get a failure on an unexpected attribute `value_parser` in the [`build_wasm_example`](https://github.com/bevyengine/bevy/blob/main/tools/build-wasm-example/src/main.rs#L9) but that should work since the `Cargo.toml` has the derive feature for `clap` (and the GitHub Action CI passes). Any idea what could be going on there?

mockersf commented:
you should probably run `cargo update` 🙂

mlodato517 commented:
I'll give that a try! If that works then does that mean we should update the version in the `Cargo.toml` because you locally need a version higher than 3.1.12?

mlodato517 commented:
Thanks, that did it! Should I toss up a PR updating the version to 3.2.x? Or is that an incorrect change?

mockersf commented:
yup please do 👍

mlodato517 commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -68,9 +76,60 @@ impl VisibleEntities {
+            let entities = self.entities_with_mesh.get_mut(&mesh)?;
+
+            // There could be duplicate entries in here if an entity was registered with a mesh
+            // twice. It's important to remove all references so that if an entity gets a new mesh
```
I wanted this change to be consistent with the comment in `register` but I didn't save the file. I'll commit this locally because this causes a reflow of the comment but I'll wait until things are approved before worrying to save on CI cycles and notifications
```suggestion
            // multiple times. It's important to remove all references so that if an entity gets a new mesh
```

mlodato517 commented:
Okay I think the duplicate-entity-mesh-relationship-bug is handled in the newest commit. I'll let that stew for a bit and clean that up and merge tomorrow if there are no new comments.

mlodato517 commented:
Rebased, squashed, and fixed up all the commits :crossed_fingers:

mlodato517 commented:
Except I forgot a silly little fix so I amended that...

superdump commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -70,9 +78,61 @@ impl VisibleEntities {
+            // multiple times. It's important to remove all references so that if an entity gets a
+            // new mesh and its old mesh is mutated, the entity doesn't get its old mesh's new
+            // `Aabb`.
+            let idxs_to_remove: SmallVec<[usize; 1]> = entities
+                .iter()
+                .enumerate()
+                .filter_map(|(i, e)| (e == &entity).then(|| i))
+                .collect();
+            let num_removed = idxs_to_remove.len();
+            for idx in idxs_to_remove.into_iter().rev() {
+                entities.swap_remove(idx);
+            }
+            Some(num_removed)
```
It's minor but I think you could just iterate the indices in reverse and swap_remove if the entity matches.

mlodato517 replied:
Ah yeah good point!

I did this so I could grab the `len()` but I guess I could just increment a counter? I wanted to record the length because it's an indication of the state of things being "wrong". In my head the actual return type of `deregister` could be:
```rust
enum SomeError {
    EntityInOneMapButNotTheOther,
    EntityRegisteredWithMeshMultipleTimes(usize),
}
fn deregister() -> Option<Result<(), SomeError>> {}
```
Though technically it's probably not useful to know how many times the entity was registered, just that it was more than once.

For this PR do you think I should:
1. leave it as is
2. don't collect and keep track of the number of items so we can return it/log it/something
3. don't collect and keep a bool to see if we removed the entity more than once with a comment indicating something would be wrong in that case
4. don't collect and don't keep track of anything

superdump replied:
As mentioned, I won’t block on it so 1 for this PR. But a few things to consider for the other answers in a separate PR or this one if you feel like it:

- Is the error a bug in bevy?
- Is it a bug in user code?
- Is the number of items removed necessary to identify that the error occurred?
- Does seeing the number of items removed help with debugging?
- What is simple to implement and understand?

mlodato517 replied:
Ah no I can't trivially iterate the indices without collecting because the iterator over the indices immutably borrows `entities` which is what I'm trying to remove from. Unless I misunderstood your comment!

mlodato517 replied:
Oh wait I probably can't do that for the same reason - the reverse iterator over the entities would borrow things when trying to remove?

mlodato517 replied:
Sorry for the spam, the marbles finally clicked together, I'm guessing you meant something like:
```rust
            for i in (0..entities.len()).rev() {
                if entities[i] == entity {
                    entities.swap_remove(i);
                }
            }
```

mlodato517 replied:
Updated in the latest commit (0a2ac7c2678931bb8e47b2a0248392e6d7848d11). I also removed the count for now because :shrug: but left some comments indicating to future devs what things might indicate problems.

superdump replied:
Yup, that was what I meant. :)

superdump commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -85,55 +145,115 @@ impl Plugin for VisibilityPlugin {
+///
+/// To opt out of bound calculation for an `Entity`, give it the [`NoAabbUpdate`] component.
+///
+/// **Note** This system needs to remove entities from their collection in
```
```suggestion
/// NOTE: This system needs to remove entities from their collection in
```

mlodato517 commented on crates/bevy_render/src/view/visibility/mod.rs:
```diff
@@ -157,60 +215,120 @@ impl Plugin for VisibilityPlugin {
+                CoreStage::PostUpdate,
+                calculate_bounds.label(CalculateBounds),
+            )
+            .add_system_to_stage(CoreStage::PostUpdate, update_bounds.label(UpdateBounds))
```
Excellent catch! Added in 9b36e77c8ab5b204c446db6c639883fdb68bf5ac.
